autoprotocol/protocol.py

```diff
@@ -5135,10 +5141,31 @@ def provision(self, resource_id, dests, volumes):
                 self.instructions[-1].to.append(xfer)
             else:
                 provision_instructions_to_return.append(
-                    self._append_and_return(Provision(resource_id, dest_group))
+                    self._append_and_return(
+                        Provision(resource_id, dest_group, measurement_mode)
+                    )
                 )
         return provision_instructions_to_return
 
+    def _identify_measurement_mode(self, volumes):
+        unique_measure_modes = set()
+        for amount in volumes:
+            print(amount.dimensionality)
+            if not isinstance(amount, (str, Unit)):
+                raise TypeError(f"Provided amount {amount} is not supported.")
+            if amount.dimensionality == Unit(1, "liter").dimensionality:
+                unique_measure_modes.add("volume")
+            elif amount.dimensionality == Unit(1, "gram").dimensionality:
+                unique_measure_modes.add("mass")
+            else:
+                raise ValueError(
+                    f"Provisioning of resources with measurement unit of {amount.unit} is not supported."
+                )
+        if len(unique_measure_modes) != 1:
+            raise ValueError("Received amounts with more than one measurement mode")
```

Comment:
Up until this line, this method has been instruction agnostic. ie, you could use `_identify_measurement_mode` to decipher a measurement mode from any list of `Unit`s provided.

Starting with this `else` block, we raise a `ValueError` based on the `provision` method's support of a type of unit. I think we should make this method clearer, maybe renaming to `_identify_provision_mode`, so it's clear this is a provision-specific function

Reply:
Renamed to `_identify_provision_mode`